File: src/main/java/com/github/tammo/fun/frontend/type/Type.java
package com.github.tammo.fun.frontend.type;

public sealed interface Type {

    static Type fromString(String type) {
        return switch (type) {
            case "Int" -> Types.Int;
            case "Boolean" -> Types.Boolean;
            case "String" -> Types.String;
            case "String[]" -> Types.StringArray;
            case "Unit" -> Types.Unit;
            default -> new TypeVariable();
        };
    }

    enum Types implements Type {
        Int,
        Boolean,
        String,
        Unit,
        StringArray
    }

    record FunctionType(Type parameter, Type returnType) implements Type {

    }

    record TypeVariable() implements Type {
    }

}

File: src/main/scala/com/github/tammo/frontend/type/Type.scala
package com.github.tammo.frontend.`type`

sealed trait Type

object Type {

  def fromString(name: String): Type = name match
    case "Int" => Int
    case "Boolean" => Boolean
    case "String" => String
    case "String[]" => StringArray
    case "Unit" => Unit
    case _ => Variable()

  sealed trait ResolvedType extends Type

  sealed trait PrimitiveType extends ResolvedType

  case object Int extends PrimitiveType

  case object String extends PrimitiveType

  case object Boolean extends PrimitiveType

  case object Unit extends PrimitiveType // TODO extends ????

  case object StringArray extends PrimitiveType

  case class FunctionType(parameter: Type, returnType: Type)
      extends ResolvedType

  case class Variable() extends Type
}
